extensions/aris/src/bigDataCluster/tree/controllerTreeNode.ts

```diff
@@ -0,0 +1,475 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+'use strict';
+
+import * as vscode from 'vscode';
+import * as azdata from 'azdata';
+import * as nls from 'vscode-nls';
+import { IControllerTreeChangeHandler } from './controllerTreeChangeHandler';
+import { TreeNode } from './treeNode';
+import { IEndPoint, IControllerError } from '../controller/types';
+import { ArisController } from '../controller/controller';
+import { IconPath, BdcItemType } from '../constants';
+
+const localize = nls.loadMessageBundle();
+
+export abstract class ControllerTreeNode extends TreeNode {
+	private _description: string;
+	private _nodeType: string;
+	private _iconPath: { dark: string, light: string };
+	private _treeChangeHandler: IControllerTreeChangeHandler;
+
+	constructor(p?: {
+		label: string;
+		treeChangeHandler: IControllerTreeChangeHandler;
+		parent?: ControllerTreeNode;
+		description?: string;
+		nodeType?: string;
+		iconPath?: { dark: string, light: string };
+	}) {
+		super();
+		this.initialize(p);
+	}
+
+	public initialize(p?: {
```

Comment:
This is a very strange pattern.
- Please just use regular constructor with individual inputs instead of these object defintions
- please don't put constructor logic in a separate initialize method unless it needs to be there (e.g. you're doing some kind of actual logic). All of this is either "pass variable to parent, where it's set as a protected field" or "define the field as a `private _treeChangeHandler` for example in this constructor.

The result of this pattern is way more lines of code and risk of missing calling this initialize method on parents. Let's avoid.

Reply:
You can initialize the object even after constructing. Java use this pattern sometimes. Fixed this anyway since no longer used.